Could not get empty comment text

<!--
Oh no, a bug! It happens. Thanks for reporting an issue with NewPipe. If this is your first bug report, read the following information before proceeding:

Please note, we only support the latest version of NewPipe. In order to check your app version, open the left drawer and click on "About". If you don't have the latest version, upgrade to it and reproduce the problem before opening the issue. The release page (https://github.com/TeamNewPipe/NewPipe/releases/latest) is where you can get it.

P.S.: Our contribution guidelines might be a nice document to read before you fill out the report :) You can find it at https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md

To make it easier for us to help you please enter detailed information in the template we have provided below. If a section isn't relevant, just delete it, though it would be helpful to still provide as much detail as possible.
-->

<!-- The comments between these brackets won't show up in the submitted issue (as you can see in the preview). -->

### Version
<!-- Which version are you using? Hopefully the latest! We just told you that above! -->
- 0.19.8

### Steps to reproduce the bug
<!--
1. Go to '...'
2. Press on '....'
3. Swipe down to '....'
-->
This can be reproduced just by opening this random stream in the logs, either sharing the link to NewPipe or pasting it in the search bar.

<!-- If you can't cause the bug to show up again reliably (and hence don't have a proper set of steps to give us), please still try to give as many details as possible on how you think you encountered the bug. -->

### Expected behavior
<!-- Tell us what you expect to happen. -->
All comments to normally load.

### Actual behaviour
<!-- Tell us what happens instead. -->
Error shows up. 
Comments are loaded but some of them maybe are missing.

### Logs
<!-- If your bug includes a crash (where you're shown the Error Report page with a bunch of info), copy it to the clipboard (there is a share button for this), head over to our bug report to markdown converter at https://teamnewpipe.github.io/CrashReportToMarkdown/ and paste it. Copy the converted text (it is MUCH easier to read this way) from there and paste it here: -->

## Exception
* __User Action:__ requested comments
* __Request:__ https://www.youtube.com/watch?v=VM_6n762j6M
* __Content Country:__ DE
* __Content Language:__ de
* __App Language:__ de_DE
* __Service:__ YouTube
* __Version:__ 0.19.8
* __OS:__ Linux Android 10 - 29
<details><summary><b>Crash log </b></summary><p>

```
org.schabi.newpipe.extractor.exceptions.ParsingException: Could not get comment text
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeCommentsInfoItemExtractor.getCommentText(YoutubeCommentsInfoItemExtractor.java:89)
	at org.schabi.newpipe.extractor.comments.CommentsInfoItemsCollector.extract(CommentsInfoItemsCollector.java:33)
	at org.schabi.newpipe.extractor.comments.CommentsInfoItemsCollector.commit(CommentsInfoItemsCollector.java:79)
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeCommentsExtractor.collectCommentsFrom(YoutubeCommentsExtractor.java:122)
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeCommentsExtractor.getPage(YoutubeCommentsExtractor.java:100)
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeCommentsExtractor.getInitialPage(YoutubeCommentsExtractor.java:52)
	at org.schabi.newpipe.extractor.utils.ExtractorHelper.getItemsPageOrLogError(ExtractorHelper.java:19)
	at org.schabi.newpipe.extractor.comments.CommentsInfo.getInfo(CommentsInfo.java:40)
	at org.schabi.newpipe.extractor.comments.CommentsInfo.getInfo(CommentsInfo.java:25)
	at org.schabi.newpipe.util.ExtractorHelper.lambda$getCommentsInfo$7(ExtractorHelper.java:155)
	at org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$Xyj84g1UeNZqX9LVYNvWZsdT-kI.call(Unknown Source:4)
	at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
	at io.reactivex.Single.subscribe(Single.java:3666)
	at io.reactivex.internal.operators.single.SingleDoOnSuccess.subscribeActual(SingleDoOnSuccess.java:35)
	at io.reactivex.Single.subscribe(Single.java:3666)
	at io.reactivex.internal.operators.maybe.MaybeFromSingle.subscribeActual(MaybeFromSingle.java:41)
	at io.reactivex.Maybe.subscribe(Maybe.java:4290)
	at io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:153)
	at io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78)
	at io.reactivex.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66)
	at io.reactivex.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42)
	at io.reactivex.Flowable.subscribe(Flowable.java:14935)
	at io.reactivex.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36)
	at io.reactivex.Maybe.subscribe(Maybe.java:4290)
	at io.reactivex.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46)
	at io.reactivex.Single.subscribe(Single.java:3666)
	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:919)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.startsWith(java.lang.String)' on a null object reference
	at org.schabi.newpipe.extractor.utils.Utils.removeUTF8BOM(Utils.java:175)
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeCommentsInfoItemExtractor.getCommentText(YoutubeCommentsInfoItemExtractor.java:87)
	... 34 more

```
</details>
<hr>


<!-- That's right, here! -->


Log https://haste.tchncs.de/siyulilija.cs

--- a/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java
+++ b/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java
@@ -47,13 +47,13 @@ public class YoutubeCommentsExtractor extends CommentsExtractor {
 
     @Override
     public InfoItemsPage<CommentsInfoItem> getInitialPage() throws IOException, ExtractionException {
-        String commentsTokenInside = findValue(responseBody, "commentSectionRenderer", "}");
-        String commentsToken = findValue(commentsTokenInside, "continuation\":\"", "\"");
+        final String commentsTokenInside = findValue(responseBody, "commentSectionRenderer", "}");
+        final String commentsToken = findValue(commentsTokenInside, "continuation\":\"", "\"");
         return getPage(getNextPage(commentsToken));
     }
 
     private Page getNextPage(JsonObject ajaxJson) throws ParsingException {
-        JsonArray arr;
+        final JsonArray arr;
         try {
             arr = JsonUtils.getArray(ajaxJson, "response.continuationContents.commentSectionContinuation.continuations");
         } catch (Exception e) {
@@ -89,14 +89,14 @@ public class YoutubeCommentsExtractor extends CommentsExtractor {
             throw new IllegalArgumentException("Page doesn't contain an URL");
         }
 
-        String ajaxResponse = makeAjaxRequest(page.getUrl());
-        JsonObject ajaxJson;
+        final String ajaxResponse = makeAjaxRequest(page.getUrl());
+        final JsonObject ajaxJson;
         try {
             ajaxJson = JsonParser.array().from(ajaxResponse).getObject(1);
         } catch (Exception e) {
             throw new ParsingException("Could not parse json data for comments", e);
         }
-        CommentsInfoItemsCollector collector = new CommentsInfoItemsCollector(getServiceId());
+        final CommentsInfoItemsCollector collector = new CommentsInfoItemsCollector(getServiceId());
         collectCommentsFrom(collector, ajaxJson);
         return new InfoItemsPage<>(collector, getNextPage(ajaxJson));
     }
@@ -160,8 +160,8 @@ public class YoutubeCommentsExtractor extends CommentsExtractor {
     }
 
     private String findValue(String doc, String start, String end) {
-        int beginIndex = doc.indexOf(start) + start.length();
-        int endIndex = doc.indexOf(end, beginIndex);
+        final int beginIndex = doc.indexOf(start) + start.length();
+        final int endIndex = doc.indexOf(end, beginIndex);
         return doc.substring(beginIndex, endIndex);
     }
 }

--- a/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java
+++ b/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java
@@ -34,7 +34,7 @@ public class YoutubeCommentsInfoItemExtractor implements CommentsInfoItemExtract
     @Override
     public String getThumbnailUrl() throws ParsingException {
         try {
-            JsonArray arr = JsonUtils.getArray(json, "authorThumbnail.thumbnails");
+            final JsonArray arr = JsonUtils.getArray(json, "authorThumbnail.thumbnails");
             return JsonUtils.getString(arr.getObject(2), "url");
         } catch (Exception e) {
             throw new ParsingException("Could not get thumbnail url", e);
@@ -82,7 +82,13 @@ public class YoutubeCommentsInfoItemExtractor implements CommentsInfoItemExtract
     @Override
     public String getCommentText() throws ParsingException {
         try {
-            String commentText = getTextFromObject(JsonUtils.getObject(json, "contentText"));
+            final JsonObject contentText = JsonUtils.getObject(json, "contentText");
+            if (contentText.isEmpty()) {
+                // completely empty comments as described in
+                // https://github.com/TeamNewPipe/NewPipeExtractor/issues/380#issuecomment-668808584
+                return "";
+            }
+            final String commentText = getTextFromObject(contentText);
             // youtube adds U+FEFF in some comments. eg. https://www.youtube.com/watch?v=Nj4F63E59io<feff>
             return Utils.removeUTF8BOM(commentText);
         } catch (Exception e) {

--- a/extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/YoutubeCommentsExtractorTest.java
+++ b/extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/YoutubeCommentsExtractorTest.java
@@ -23,91 +23,140 @@ import static org.junit.Assert.assertTrue;
 import static org.schabi.newpipe.extractor.ServiceList.YouTube;
 
 public class YoutubeCommentsExtractorTest {
-    private static final String urlYT = "https://www.youtube.com/watch?v=D00Au7k3i6o";
-    private static final String urlInvidious = "https://invidio.us/watch?v=D00Au7k3i6o";
-    private static YoutubeCommentsExtractor extractorYT;
-    private static YoutubeCommentsExtractor extractorInvidious;
+    /**
+     * Test a "normal" YouTube and Invidious page
+     */
+    public static class Thomas {
+        private static final String urlYT = "https://www.youtube.com/watch?v=D00Au7k3i6o";
+        private static final String urlInvidious = "https://invidio.us/watch?v=D00Au7k3i6o";
+        private static YoutubeCommentsExtractor extractorYT;
+        private static YoutubeCommentsExtractor extractorInvidious;
 
-    @BeforeClass
-    public static void setUp() throws Exception {
-        NewPipe.init(DownloaderTestImpl.getInstance());
-        extractorYT = (YoutubeCommentsExtractor) YouTube
-                .getCommentsExtractor(urlYT);
-        extractorYT.fetchPage();
-        extractorInvidious = (YoutubeCommentsExtractor) YouTube
-                .getCommentsExtractor(urlInvidious);
-    }
+        private static final String commentContent = "sub 4 sub";
 
-    @Test
-    public void testGetComments() throws IOException, ExtractionException {
-        assertTrue(getCommentsHelper(extractorYT));
-        assertTrue(getCommentsHelper(extractorInvidious));
-    }
-
-    private boolean getCommentsHelper(YoutubeCommentsExtractor extractor) throws IOException, ExtractionException {
-        InfoItemsPage<CommentsInfoItem> comments = extractor.getInitialPage();
-        boolean result = findInComments(comments, "s1ck m3m3");
-
-        while (comments.hasNextPage() && !result) {
-            comments = extractor.getPage(comments.getNextPage());
-            result = findInComments(comments, "s1ck m3m3");
+        @BeforeClass
+        public static void setUp() throws Exception {
+            NewPipe.init(DownloaderTestImpl.getInstance());
+            extractorYT = (YoutubeCommentsExtractor) YouTube
+                    .getCommentsExtractor(urlYT);
+            extractorYT.fetchPage();
+            extractorInvidious = (YoutubeCommentsExtractor) YouTube
+                    .getCommentsExtractor(urlInvidious);
+            extractorInvidious.fetchPage();
         }
 
-        return result;
-    }
-
-    @Test
-    public void testGetCommentsFromCommentsInfo() throws IOException, ExtractionException {
-        assertTrue(getCommentsFromCommentsInfoHelper(urlYT));
-        assertTrue(getCommentsFromCommentsInfoHelper(urlInvidious));
-    }
-
-    private boolean getCommentsFromCommentsInfoHelper(String url) throws IOException, ExtractionException {
-        CommentsInfo commentsInfo = CommentsInfo.getInfo(url);
-
-        assertEquals("Comments", commentsInfo.getName());
-        boolean result = findInComments(commentsInfo.getRelatedItems(), "s1ck m3m3");
-
-        Page nextPage = commentsInfo.getNextPage();
-        InfoItemsPage<CommentsInfoItem> moreItems = new InfoItemsPage<>(null, nextPage, null);
-        while (moreItems.hasNextPage() && !result) {
-            moreItems = CommentsInfo.getMoreItems(YouTube, commentsInfo, nextPage);
-            result = findInComments(moreItems.getItems(), "s1ck m3m3");
-            nextPage = moreItems.getNextPage();
+        @Test
+        public void testGetComments() throws IOException, ExtractionException {
+            assertTrue(getCommentsHelper(extractorYT));
+            assertTrue(getCommentsHelper(extractorInvidious));
         }
-        return result;
-    }
 
-    @Test
-    public void testGetCommentsAllData() throws IOException, ExtractionException {
-        InfoItemsPage<CommentsInfoItem> comments = extractorYT.getInitialPage();
+        private boolean getCommentsHelper(YoutubeCommentsExtractor extractor) throws IOException, ExtractionException {
+            InfoItemsPage<CommentsInfoItem> comments = extractor.getInitialPage();
+            boolean result = findInComments(comments, commentContent);
 
-        DefaultTests.defaultTestListOfItems(YouTube, comments.getItems(), comments.getErrors());
-        for (CommentsInfoItem c : comments.getItems()) {
-            assertFalse(Utils.isBlank(c.getUploaderUrl()));
-            assertFalse(Utils.isBlank(c.getUploaderName()));
-            assertFalse(Utils.isBlank(c.getUploaderAvatarUrl()));
-            assertFalse(Utils.isBlank(c.getCommentId()));
-            assertFalse(Utils.isBlank(c.getCommentText()));
-            assertFalse(Utils.isBlank(c.getName()));
-            assertFalse(Utils.isBlank(c.getTextualUploadDate()));
-            assertNotNull(c.getUploadDate());
-            assertFalse(Utils.isBlank(c.getThumbnailUrl()));
-            assertFalse(Utils.isBlank(c.getUrl()));
-            assertFalse(c.getLikeCount() < 0);
+            while (comments.hasNextPage() && !result) {
+                comments = extractor.getPage(comments.getNextPage());
+                result = findInComments(comments, commentContent);
+            }
+
+            return result;
         }
-    }
 
-    private boolean findInComments(InfoItemsPage<CommentsInfoItem> comments, String comment) {
-        return findInComments(comments.getItems(), comment);
-    }
+        @Test
+        public void testGetCommentsFromCommentsInfo() throws IOException, ExtractionException {
+            assertTrue(getCommentsFromCommentsInfoHelper(urlYT));
+            assertTrue(getCommentsFromCommentsInfoHelper(urlInvidious));
+        }
 
-    private boolean findInComments(List<CommentsInfoItem> comments, String comment) {
-        for (CommentsInfoItem c : comments) {
-            if (c.getCommentText().contains(comment)) {
-                return true;
+        private boolean getCommentsFromCommentsInfoHelper(String url) throws IOException, ExtractionException {
+            final CommentsInfo commentsInfo = CommentsInfo.getInfo(url);
+
+            assertEquals("Comments", commentsInfo.getName());
+            boolean result = findInComments(commentsInfo.getRelatedItems(), commentContent);
+
+            Page nextPage = commentsInfo.getNextPage();
+            InfoItemsPage<CommentsInfoItem> moreItems = new InfoItemsPage<>(null, nextPage, null);
+            while (moreItems.hasNextPage() && !result) {
+                moreItems = CommentsInfo.getMoreItems(YouTube, commentsInfo, nextPage);
+                result = findInComments(moreItems.getItems(), commentContent);
+                nextPage = moreItems.getNextPage();
+            }
+            return result;
+        }
+
+        @Test
+        public void testGetCommentsAllData() throws IOException, ExtractionException {
+            InfoItemsPage<CommentsInfoItem> comments = extractorYT.getInitialPage();
+
+            DefaultTests.defaultTestListOfItems(YouTube, comments.getItems(), comments.getErrors());
+            for (CommentsInfoItem c : comments.getItems()) {
+                assertFalse(Utils.isBlank(c.getUploaderUrl()));
+                assertFalse(Utils.isBlank(c.getUploaderName()));
+                assertFalse(Utils.isBlank(c.getUploaderAvatarUrl()));
+                assertFalse(Utils.isBlank(c.getCommentId()));
+                assertFalse(Utils.isBlank(c.getCommentText()));
+                assertFalse(Utils.isBlank(c.getName()));
+                assertFalse(Utils.isBlank(c.getTextualUploadDate()));
+                assertNotNull(c.getUploadDate());
+                assertFalse(Utils.isBlank(c.getThumbnailUrl()));
+                assertFalse(Utils.isBlank(c.getUrl()));
+                assertFalse(c.getLikeCount() < 0);
             }
         }
-        return false;
+
+        private boolean findInComments(InfoItemsPage<CommentsInfoItem> comments, String comment) {
+            return findInComments(comments.getItems(), comment);
+        }
+
+        private boolean findInComments(List<CommentsInfoItem> comments, String comment) {
+            for (CommentsInfoItem c : comments) {
+                if (c.getCommentText().contains(comment)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Test a video with an empty comment
+     */
+    public static class EmptyComment {
+        private static YoutubeCommentsExtractor extractor;
+        private final static String url = "https://www.youtube.com/watch?v=VM_6n762j6M";
+
+        @BeforeClass
+        public static void setUp() throws Exception {
+            NewPipe.init(DownloaderTestImpl.getInstance());
+            extractor = (YoutubeCommentsExtractor) YouTube
+                    .getCommentsExtractor(url);
+            extractor.fetchPage();
+        }
+
+        @Test
+        public void testGetCommentsAllData() throws IOException, ExtractionException {
+            final InfoItemsPage<CommentsInfoItem> comments = extractor.getInitialPage();
+
+            DefaultTests.defaultTestListOfItems(YouTube, comments.getItems(), comments.getErrors());
+            for (CommentsInfoItem c : comments.getItems()) {
+                assertFalse(Utils.isBlank(c.getUploaderUrl()));
+                assertFalse(Utils.isBlank(c.getUploaderName()));
+                assertFalse(Utils.isBlank(c.getUploaderAvatarUrl()));
+                assertFalse(Utils.isBlank(c.getCommentId()));
+                assertFalse(Utils.isBlank(c.getName()));
+                assertFalse(Utils.isBlank(c.getTextualUploadDate()));
+                assertNotNull(c.getUploadDate());
+                assertFalse(Utils.isBlank(c.getThumbnailUrl()));
+                assertFalse(Utils.isBlank(c.getUrl()));
+                assertFalse(c.getLikeCount() < 0);
+                if (c.getCommentId().equals("Ugga_h1-EXdHB3gCoAEC")) { // comment without text
+                    assertTrue(Utils.isBlank(c.getCommentText()));
+                } else {
+                    assertFalse(Utils.isBlank(c.getCommentText()));
+                }
+            }
+        }
+
     }
 }

